gflownet/envs/ctorus.py

```diff
@@ -258,22 +393,64 @@ def step(
             False, if the action is not allowed for the current state, e.g. stop at the
             root state
         """
+        # If done is True, return invalid
         if self.done:
             return self.state, action, False
-        # If only possible action is eos, then force eos
-        # If the number of actions is equal to maximum trajectory length
-        elif self.n_actions == self.length_traj:
+        # If action is EOS, check that the number of steps is equal to the trajectory
+        # length, set done to True, increment n_actions and return same state
+        elif action == self.eos:
+            assert self.state[-1] == self.length_traj
             self.done = True
             self.n_actions += 1
             return self.state, self.eos, True
-        # If action is eos, then it is invalid
-        elif action == self.eos:
-            return self.state, action, False
         # Otherwise perform action
         else:
             self.n_actions += 1
-            for dim, angle in enumerate(action):
-                self.state[int(dim)] += angle
-                self.state[int(dim)] = self.state[int(dim)] % (2 * np.pi)
-                self.state[-1] = self.n_actions
+            self._step(action, backward=False)
+            return self.state, action, True
+
+    def step_backwards(
+        self, action: Tuple[float], skip_mask_check: bool = False
+    ) -> Tuple[List[float], Tuple[int, float], bool]:
+        """
+        Executes backward step given an action.
+
+        See: _step().
+
+        Args
+        ----
+        action : tuple
+            Action to be executed. An action is a vector where the value at position d
+            indicates the increment in the angle at dimension d.
+
+        skip_mask_check : bool
+            Ignored because the action space space is fully continuous, therefore there
+            is nothing to check.
+
+        Returns
+        -------
+        self.state : list
+            The sequence after executing the action
+
+        action : int
+            Action executed
+
+        valid : bool
+            False, if the action is not allowed for the current state, e.g. stop at the
+            root state
+        """
+        # If done is True, set done to False, increment n_actions and return same state
+        if self.done:
+            assert action == self.eos
+            self.done = False
+            self.n_actions += 1
```

Comment:
Why `self.n_actions` is incremented in the backward step, not decremented? I thought it should be the same as the last dimension of the state

Reply:
No, it is not the same. `self.n_actions` counts the number of (valid) actions in a trajectory, regardless of whether it is forward of backward.